File: Backend/rate_limit.py
"""
Lightweight in-memory rate limiter (§5.6).

A sliding-window counter keyed on (scope, client-IP), exposed as a FastAPI
dependency so expensive/abusable routes (LLM calls, auth) can't be hammered for
financial-DoS or brute-force.

NOTE: state is per-process. Under multiple uvicorn workers each has its own
buckets, so the effective limit is (limit × workers). Move the store to Redis when
scaling beyond one process (pairs with the §10.9 queue/broker work).
"""

import time
from collections import defaultdict, deque

from fastapi import Request, HTTPException

_BUCKETS: dict[str, deque] = defaultdict(deque)


class RateLimiter:
    def __init__(self, max_calls: int, window_seconds: int, scope: str = ""):
        self.max_calls = max_calls
        self.window = window_seconds
        self.scope = scope

    def __call__(self, request: Request):
        ip = request.client.host if request.client else "unknown"
        key = f"{self.scope or request.url.path}:{ip}"
        now = time.monotonic()
        dq = _BUCKETS[key]

        cutoff = now - self.window
        while dq and dq[0] <= cutoff:
            dq.popleft()

        if len(dq) >= self.max_calls:
            retry = int(self.window - (now - dq[0])) + 1
            raise HTTPException(
                status_code=429,
                detail="Too many requests — please slow down.",
                headers={"Retry-After": str(retry)},
            )

        dq.append(now)

File: Satellite/middleware/rateLimit.js
// middleware/rateLimit.js — lightweight in-memory rate limiter (§5.6)
//
// Sliding-window counter keyed on (scope, client-IP), so expensive/abusable routes
// (LLM generation) can't be hammered for financial-DoS. NOTE: per-process state —
// move to Redis when running more than one Satellite instance (§10.9).

const buckets = new Map();

/**
 * @param {number} maxCalls  max requests allowed within the window
 * @param {number} windowMs  window size in milliseconds
 * @param {string} scope     bucket namespace (defaults to the route path)
 */
function rateLimit(maxCalls, windowMs, scope = "") {
  return (req, res, next) => {
    const ip = req.ip || (req.connection && req.connection.remoteAddress) || "unknown";
    const key = `${scope || req.baseUrl + req.path}:${ip}`;
    const now = Date.now();

    let dq = buckets.get(key);
    if (!dq) {
      dq = [];
      buckets.set(key, dq);
    }

    const cutoff = now - windowMs;
    while (dq.length && dq[0] <= cutoff) dq.shift();

    if (dq.length >= maxCalls) {
      const retry = Math.ceil((dq[0] + windowMs - now) / 1000);
      res.set("Retry-After", String(retry));
      return res.status(429).json({ error: "Too many requests — please slow down." });
    }

    dq.push(now);
    next();
  };
}

module.exports = { rateLimit };
